src/modules/networks/networks-store.ts

```diff
@@ -1,101 +1,171 @@
+    );
+    const updatedNetworkConfigs = mergeNetworkConfigs(
+      commonNetworkConfis.status === 'fulfilled'
+        ? commonNetworkConfis.value
+        : [],
+      extraNetworkConfigs.status === 'fulfilled'
+        ? extraNetworkConfigs.value
+        : []
+    );
+
+    this.networkConfigs = mergeNetworkConfigs(
+      this.networkConfigs,
+      updatedNetworkConfigs
+    );
+    this.customNetworkConfigs =
+      savedChainConfigs?.ethereumChainConfigs
+        ?.filter((config) => isCustomNetworkId(config.id))
+        .map((config) => toNetworkConfig(config.value)) || [];
+
+    return this.updateNetworks();
+  }
+
+  private async fetchNetworkById(chainId: number) {
+    const hasInformationAboutChain = [
+      ...this.networkConfigs,
+      ...this.customNetworkConfigs,
+    ].some((network) => getChainId(network) === chainId);
+    const existingNetworks = this.getState().networks;
+    if (hasInformationAboutChain && existingNetworks) {
+      return existingNetworks;
+    }
+
+    const network = await getNetworkByChainId(chainId);
+    if (network) {
+      this.networkConfigs = mergeNetworkConfigs(this.networkConfigs, [network]);
     }
-    return this.loaderPromise
-      .then(() => {
-        const { networks } = this.getState();
-        if (!networks) {
-          throw new Error('networks are expected to be not null after load()');
-        }
-        if (this.needsRetry) {
-          this.retry();
-        }
-        return networks;
-      })
-      .catch((error) => {
-        this.loaderPromise = null;
-        throw error;
-      });
+    return this.updateNetworks();
+  }
+
+  async pushConfigs(...extraNetworkConfigs: NetworkConfig[]) {
+    this.networkConfigs = mergeNetworkConfigs(
+      this.networkConfigs,
+      extraNetworkConfigs
+    );
+    return this.updateNetworks();
+  }
+
+  async load(chains?: string[]) {
+    const key = JSON.stringify(chains || []);
+    if (!this.loaderPromises[key]) {
+      this.loaderPromises[key] = this.fetchNetworks({ chains });
+    }
+    return this.loaderPromises[key].then((networks) => {
+      delete this.loaderPromises[key];
+      return networks;
+    });
   }
 
   async update() {
-    this.loaderPromise = null;
-    return this.fetchAndUpdate();
+    const key = 'update';
+    if (!this.loaderPromises[key]) {
+      this.loaderPromises[key] = this.fetchNetworks({ update: true });
+    }
+    return this.loaderPromises[key].then((networks) => {
+      delete this.loaderPromises[key];
+      return networks;
+    });
   }
 
-  retry = throttle(() => {
-    this.fetchAndUpdate();
-  }, 15000);
+  async loadNetworksWithChainId(chainId: number) {
+    const key = `chainId-${chainId}`;
+    if (!this.loaderPromises[key]) {
+      this.loaderPromises[key] = this.fetchNetworkById(chainId);
+    }
+    return this.loaderPromises[key].then((networks) => {
+      delete this.loaderPromises[key];
+      return networks;
+    });
```

Comment:
look like we're creating a new promise on each call of this method, and doing the same thing inside

Reply:
Hm. Do we? I think, it is created only if there is no current promise. Otherwise, we are setting another .then

Reply:
Yes.
Imagine 3 calls to `loadNetworksWithChainId()`. Each call will attach a new 
```
.then(() => /* delete this.loaderPromises[key]; */)
```

So when `this.fetchNetworkById(chainId);` resolves, `delete` will be called 3 times.

Also, if `loadNetworksWithChainId()` is called again after previous calls are resolved, it will call `this.fetchNetworkById(chainId);`. Maybe this part is by design.

Reply:
The second part is by design - it will stop inside the fetch function if data is exists.
So here is the same logic like in ready() for stores.

Reply:
I will think about this today. It definitely can be done better

Reply:
Oh, I remember the idea. This promises should help us to avoid parallel requests. However, it is important that all future requests will get the most up-to-date networks object. So I clear these promises after the successful call. What I should do is to clear them after errors as well

Reply:
Let's discuss this